bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxCommunicator.java

```diff
@@ -0,0 +1,521 @@
+
+        try {
+            logger.trace("CaddxCommunicator.handleBinaryProtocol() Exchanging");
+
+            if (!exchanger.offer(caddxMessage, 3, TimeUnit.SECONDS)) {
+                logger.debug("Offered message was not received");
+            }
+        } catch (InterruptedException e) {
+            logger.trace("CaddxCommunicator.handleBinaryProtocol() InterruptedException caught.");
+            Thread.currentThread().interrupt();
+        }
+
+        // Initialize state for next reception
+        inMessage = false;
+        haveMessageLength = false;
+        haveFirstByte = false;
+        messageBufferLength = 0;
+        message = new byte[0];
+        messageBufferIndex = 0;
+        unStuff = false;
+    }
+
+    private int readAsciiByte(InputStream stream) {
+        if (!haveFirstByte) { // this is the 1st digit
+            int b = readByte(in);
+            if (b == -1) {
+                return -1;
+            }
+            tempAsciiByte = (b - 0x30) * 0x10;
+            haveFirstByte = true;
+        }
+
+        if (haveFirstByte) { // this is the 2nd digit
+            int b = readByte(in);
+            if (b == -1) {
+                return -1;
+            }
+            tempAsciiByte += (b - 0x30);
+            haveFirstByte = false;
+        }
+
+        return tempAsciiByte;
+    }
+
+    private void receiveInAsciiProtocol(SerialPortEvent serialPortEvent) {
+        int b = 0;
+
+        // Read the start byte
+        if (!inMessage) // skip until 0x0A
+        {
+            while (b != 0x0A && b != -1) {
+                b = readByte(in);
+            }
+            if (b == -1) {
+                return;
+            }
+
+            inMessage = true;
+            haveMessageLength = false;
+            haveFirstByte = false;
+            messageBufferLength = 0;
+        }
+        logger.trace("CaddxCommunicator.handleAsciiProtocol() Got start byte");
+
+        // Read the message length
+        if (!haveMessageLength) {
+            while (!haveMessageLength) {
+                b = readAsciiByte(in);
+                if (b == -1) {
+                    return;
+                }
+
+                messageBufferLength = b + 2; // add 2 bytes for the checksum
+                haveMessageLength = true;
+            }
+
+            message = new byte[messageBufferLength];
+        }
+        logger.trace("CaddxCommunicator.handleAsciiProtocol() Got message length {}", b);
+
+        // Read the message
+        while (true) {
+            // Read a byte
+            while (true) {
+                b = readAsciiByte(in);
+                if (b == -1) {
+                    return;
+                }
+
+                message[messageBufferIndex] = (byte) b;
+                break;
+            }
+
+            messageBufferIndex++;
+            if (messageBufferIndex == messageBufferLength) {
+                break;
+            }
+        }
```

Comment:
can you turn this into a do-while loop instead?

Reply:
Done